File: modules/drive-sense/src/extract/vec.ts
// Three-vector arithmetic for the extraction reference. The ports implement exactly these
// definitions (notably `normalize`'s zero case and `angle` via atan2), so edge cases agree.
import { EPS } from './constants';

export type Vec3 = readonly [number, number, number];

export const ZERO: Vec3 = [0, 0, 0];

export const add = (a: Vec3, b: Vec3): Vec3 => [a[0] + b[0], a[1] + b[1], a[2] + b[2]];
export const sub = (a: Vec3, b: Vec3): Vec3 => [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
export const scale = (a: Vec3, k: number): Vec3 => [a[0] * k, a[1] * k, a[2] * k];
export const dot = (a: Vec3, b: Vec3): number => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
export const cross = (a: Vec3, b: Vec3): Vec3 => [
  a[1] * b[2] - a[2] * b[1],
  a[2] * b[0] - a[0] * b[2],
  a[0] * b[1] - a[1] * b[0],
];
export const norm = (a: Vec3): number => Math.sqrt(dot(a, a));

/** Unit vector along `a`, or exactly `[0, 0, 0]` when |a| < EPS. */
export function normalize(a: Vec3): Vec3 {
  const n = norm(a);
  return n < EPS ? ZERO : [a[0] / n, a[1] / n, a[2] / n];
}

/** The component of `v` perpendicular to the unit vector `n`: v − (v·n)n. */
export const reject = (v: Vec3, n: Vec3): Vec3 => sub(v, scale(n, dot(v, n)));

/** Angle between `a` and `b` in [0, π], as atan2(|a×b|, a·b); 0 when either is zero. */
export const angle = (a: Vec3, b: Vec3): number => Math.atan2(norm(cross(a, b)), dot(a, b));

export const isZero = (a: Vec3): boolean => a[0] === 0 && a[1] === 0 && a[2] === 0;

export const clamp = (x: number, lo: number, hi: number): number => Math.min(hi, Math.max(lo, x));

File: modules/drive-sense/android/src/main/java/expo/modules/drivesense/Vec.kt
package expo.modules.drivesense

import kotlin.math.atan2
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Three-vector arithmetic, a line-for-line port of `src/extract/vec.ts`. The edge cases (normalize's
 * exact zero, angle via atan2) must agree with the reference for the golden vectors to pass.
 */
data class Vec3(val x: Double, val y: Double, val z: Double) {
  companion object {
    val ZERO = Vec3(0.0, 0.0, 0.0)
  }
}

fun add(a: Vec3, b: Vec3): Vec3 = Vec3(a.x + b.x, a.y + b.y, a.z + b.z)

fun sub(a: Vec3, b: Vec3): Vec3 = Vec3(a.x - b.x, a.y - b.y, a.z - b.z)

fun scale(a: Vec3, k: Double): Vec3 = Vec3(a.x * k, a.y * k, a.z * k)

fun dot(a: Vec3, b: Vec3): Double = a.x * b.x + a.y * b.y + a.z * b.z

fun cross(a: Vec3, b: Vec3): Vec3 = Vec3(
  a.y * b.z - a.z * b.y,
  a.z * b.x - a.x * b.z,
  a.x * b.y - a.y * b.x
)

fun norm(a: Vec3): Double = sqrt(dot(a, a))

/** Unit vector along [a], or exactly [0, 0, 0] when |a| < EPS. */
fun normalize(a: Vec3): Vec3 {
  val n = norm(a)
  return if (n < EPS) Vec3.ZERO else Vec3(a.x / n, a.y / n, a.z / n)
}

/** v − (v·n)n */
fun reject(v: Vec3, n: Vec3): Vec3 = sub(v, scale(n, dot(v, n)))

/** atan2(|a×b|, a·b); 0 when either is zero. */
fun angle(a: Vec3, b: Vec3): Double = atan2(norm(cross(a, b)), dot(a, b))

fun isZero(a: Vec3): Boolean = a.x == 0.0 && a.y == 0.0 && a.z == 0.0

/** JS `Math.min(hi, Math.max(lo, x))`. */
fun clamp(x: Double, lo: Double, hi: Double): Double = min(hi, max(lo, x))

